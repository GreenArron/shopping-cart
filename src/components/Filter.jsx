import PropTypes from "prop-types";
import { useRef, useState } from "react";
import { useClickedAway } from "../utils/custom-hooks";
import styles from "./Filter.module.css";

function capitalizeCase(string) {
  return string.charAt(0).toUpperCase() + string.slice(1);
}

function Filter({
  name,
  currentFilter,
  setFilter,
  options,
  defaultOption,
  ...other
}) {
  const [active, setActive] = useState(false);
  const filterRef = useRef(null);
  const usedOptions =
    defaultOption === undefined ? options : [defaultOption, ...options];
  const shownName = options.includes(currentFilter)
    ? `${name}: ${currentFilter}`
    : name;

  useClickedAway(filterRef, () => {
    if (active) {
      setActive(false);
    }
  });

  function selectFilter(key) {
    setFilter(key);
    setActive(false);
  }

  return (
    <div className={styles["filter-container"]} ref={filterRef}>
      <button onClick={() => setActive((value) => !value)}>{shownName}</button>
      <div
        className={`${styles["filter-content"]} ${active && styles["float"]}`}
        {...other}
      >
        {active &&
          usedOptions.map((option) => {
            return (
              <button
                className={`${styles["filter-option"]} ${option === currentFilter && styles["selected"]}`}
                key={option}
                onClick={() => selectFilter(option)}
              >
                {capitalizeCase(option)}
              </button>
            );
          })}
      </div>
    </div>
  );
}

Filter.propTypes = {
  name: PropTypes.string.isRequired,
  currentFilter: PropTypes.string,
  setFilter: PropTypes.func,
  options: PropTypes.arrayOf(PropTypes.string),
  defaultOption: PropTypes.string,
};

export default Filter;
