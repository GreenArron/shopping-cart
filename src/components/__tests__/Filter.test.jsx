import { it, expect, vi } from "vitest";
import { render, screen } from "@testing-library/react";
import UserEvent from "@testing-library/user-event";
import Filter from "../Filter";

const mockedFilter = (fn) => (
  <Filter
    name="testfilter"
    defaultOption={"testdefault"}
    currentFilter={"testselected"}
    options={["testselected"]}
    setFilter={fn}
  />
);

it("matches the last approved snapshots", async () => {
  const user = UserEvent.setup();
  const { container } = render(mockedFilter());
  expect(container).toMatchSnapshot();
  await user.click(screen.getByRole("button"));
  expect(container).toMatchSnapshot();
});

it("sets filter on click", async () => {
  const user = UserEvent.setup();
  const callback = vi.fn();
  render(mockedFilter(callback));
  await user.click(screen.getByRole("button"));
  await user.click(screen.getByRole("button", { name: /testdefault/i }));
  expect(callback).toHaveBeenCalledOnce();
});
