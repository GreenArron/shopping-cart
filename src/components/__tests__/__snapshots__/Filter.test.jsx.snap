// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`matches the last approved snapshots 1`] = `
<div>
  <div
    class="_filter-container_e8fa16"
  >
    <button>
      testfilter: testselected
    </button>
    <div
      class="_filter-content_e8fa16 false"
    />
  </div>
</div>
`;

exports[`matches the last approved snapshots 2`] = `
<div>
  <div
    class="_filter-container_e8fa16"
  >
    <button>
      testfilter: testselected
    </button>
    <div
      class="_filter-content_e8fa16 _float_e8fa16"
    >
      <button
        class="_filter-option_e8fa16 false"
      >
        Testdefault
      </button>
      <button
        class="_filter-option_e8fa16 _selected_e8fa16"
      >
        Testselected
      </button>
    </div>
  </div>
</div>
`;
